### ISSUE TITLE
AMD MI300A Unified Memory Support

### ISSUE THREAD

**Issue by lancelotnd** [pytorch/pytorch#145693]
### 🚀 The feature, motivation and pitch

I am working on improving performance for LLM workloads on AMD Instinct™ MI300A Accelerator. This APU is equipped with a fully unified-memory architecture not taken advantage by PyTorch at this time. Because both the GPU and CPU share the same physical memory, memcpy ops become redundant and can limit the size of the model we can train. Adding support for unified memory on ROCm for that particular APU  would allow for zero-copy operations.

The motivation is similar that of #140787, but for ROCm instead of MPS.

Given that this APU is targeted for the most demanding HPC ML workloads, there is a great interest in optimizing the performance of PyTorch for it. Notably, El Capitan, the top1 Supercomputer from [top500](https://top500.org/) runs exclusively with AMD's MI300A.

### Additional context

[MI300a Data Sheet](https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/data-sheets/amd-instinct-mi300a-data-sheet.pdf)

**Comment by jayfurmanek** [comment]
We did have [an RFC](https://github.com/pytorch/rfcs/pull/36) for this very thing a while back. The idea is to provide zero-copy for tensors across CPU/GPU but the usage model in torch has implicit and explicit copies that have to be handled properly, which is a bit tricky. Unified addressing also doesn't prevent you from having to synchronize. So the question becomes can we safely make these changes and get a performance win with a framework that does a pretty good job avoiding copy latency in the first place.

**Comment by lancelotnd** [comment]
Thank you very much for resurfacing this RFC. I really appreciate all the work that has been done toward the design of this implementation. I'll start off of that document to build a prototype and test it on MI300A nodes. And as tricky as it may be, I think it is still worth the effort now that unified memory is no longer just virtual.

**Comment by malfet** [comment]
We can merge this into a more generic issue about unified memory support. Also, I don't know how/if unified memory is indeed supported by ROCm SW stack
